src/wallet/wallet.cpp

```diff
@@ -35,6 +35,8 @@
 #include <boost/algorithm/string/replace.hpp>
 #include <boost/thread.hpp>
 
+/** Protects the vpwallets vector */
+CCriticalSection cs_wallets;
 std::vector<CWalletRef> vpwallets;
```

Comment:
I had the idea that `walletinit.h/.c` is kind of the multiwallet interface (wallet manager). The functions we have there (`DeleteWallets()`, `VerifyWallets()`, `ShutdownWallets();`, etc.) sounded for me after `vpwallets` belongs there... 

IMO, the multiwallet map should ideally not sit within the objects (`CWallet`) implementation.